C++ Support doesn't work

# Problem

When trying to use Zephyr with C++ I noticed that a lot of the headers contain code that isn't valid in C++ and causes compile errors when included in C++ files. I noticed this mainly with the Bluetooth subsystem but I wouldn't be surprised if other parts of are affected too as I haven't tested them.

So even though my examples focus on Bluetooth this issue concerns Zephyr as a whole.

The documentation claims that Zephyr has C++ support, but at the moment not even the most basic things work.

After asking on IRC about this @SebastianBoe confirmed those problems and submitted PR #5642 with some small fixes. While that was a first step in the right direction a lot more has to be done still.

As far as I know there are absolutely no automated tests in Zephyr for C++ compatibility at the moment. I would suggest changing that to properly ensure C++ compatibility in the future. If possible maybe just reuse/adopt existing C tests and additionally compile them in C++.

# Solution

1) Fix any headers that contain invalid C++ code to be valid in both C and C++.
2) Add automated tests for C++ compatibility

# Minimal Example

```C++
#include <zephyr.h>

#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/conn.h>
#include <bluetooth/uuid.h>
#include <bluetooth/gatt.h>

static const struct bt_data ad[] = {
    BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR))
};

void on_bt_ready(int err) {
    bl_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), NULL, NULL);
}

int main() {
    bt_enable(on_bt_ready);
}
```

The following configuration was used in `prj.conf`:

```
CONFIG_CPLUSPLUS=y
CONFIG_BT=y
CONFIG_BT_DEBUG_LOG=y
CONFIG_BT_DEVICE_NAME="Test beacon"
```

Built with the current master branch of zephyr and zephyr-sdk-0.9.2 for the nrf52 platform using the following commands:

```bash
mkdir build && cd build
cmake -DBOARD=nrf52_pca10040 ..
make
```

This minimal example will generate the following errors:

```
In file included from /home/zigarrre/coding/zephyr_test/src/test.cpp:3:0:
/home/zigarrre/coding/zephyr_test/src/test.cpp: In function ‘void on_bt_ready(int)’:
/home/zigarrre/coding/z/zephyr/include/bluetooth/bluetooth.h:149:4: error: taking address of temporary [-fpermissive]
    })
    ^
/home/zigarrre/coding/z/zephyr/include/bluetooth/bluetooth.h:151:24: note: in expansion of macro ‘BT_LE_ADV_PARAM’
 #define BT_LE_ADV_CONN BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE, \
                        ^~~~~~~~~~~~~~~
/home/zigarrre/coding/zephyr_test/src/test.cpp:14:21: note: in expansion of macro ‘BT_LE_ADV_CONN’
     bl_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), NULL, NULL);
                     ^~~~~~~~~~~~~~
In file included from /home/zigarrre/coding/z/zephyr/include/kernel.h:27:0,
                 from /home/zigarrre/coding/z/zephyr/include/zephyr.h:18,
                 from /home/zigarrre/coding/zephyr_test/src/test.cpp:1:
/home/zigarrre/coding/z/zephyr/include/misc/util.h:42:35: error: ‘__builtin_types_compatible_p’ was not declared in this scope
            __typeof__(&(array)[0])))
                                   ^
/home/zigarrre/coding/z/zephyr/include/misc/util.h:34:66: note: in definition of macro ‘ZERO_OR_COMPILE_ERROR’
 #define ZERO_OR_COMPILE_ERROR(cond) ((int) sizeof(char[1 - 2 * !(cond)]) - 1)
                                                                  ^~~~
/home/zigarrre/coding/z/zephyr/include/misc/util.h:48:20: note: in expansion of macro ‘IS_ARRAY’
  ((unsigned long) (IS_ARRAY(array) + \
                    ^~~~~~~~
/home/zigarrre/coding/zephyr_test/src/test.cpp:14:41: note: in expansion of macro ‘ARRAY_SIZE’
     bl_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), NULL, NULL);
                                         ^~~~~~~~~~
/home/zigarrre/coding/z/zephyr/include/misc/util.h:34:51: error: expected primary-expression before ‘char’
 #define ZERO_OR_COMPILE_ERROR(cond) ((int) sizeof(char[1 - 2 * !(cond)]) - 1)
                                                   ^
/home/zigarrre/coding/z/zephyr/include/misc/util.h:40:2: note: in expansion of macro ‘ZERO_OR_COMPILE_ERROR’
  ZERO_OR_COMPILE_ERROR( \
  ^~~~~~~~~~~~~~~~~~~~~
/home/zigarrre/coding/z/zephyr/include/misc/util.h:48:20: note: in expansion of macro ‘IS_ARRAY’
  ((unsigned long) (IS_ARRAY(array) + \
                    ^~~~~~~~
/home/zigarrre/coding/zephyr_test/src/test.cpp:14:41: note: in expansion of macro ‘ARRAY_SIZE’
     bl_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), NULL, NULL);
                                         ^~~~~~~~~~
/home/zigarrre/coding/z/zephyr/include/misc/util.h:34:51: error: expected ‘)’ before ‘char’
 #define ZERO_OR_COMPILE_ERROR(cond) ((int) sizeof(char[1 - 2 * !(cond)]) - 1)
                                                   ^
/home/zigarrre/coding/z/zephyr/include/misc/util.h:40:2: note: in expansion of macro ‘ZERO_OR_COMPILE_ERROR’
  ZERO_OR_COMPILE_ERROR( \
  ^~~~~~~~~~~~~~~~~~~~~
/home/zigarrre/coding/z/zephyr/include/misc/util.h:48:20: note: in expansion of macro ‘IS_ARRAY’
  ((unsigned long) (IS_ARRAY(array) + \
                    ^~~~~~~~
/home/zigarrre/coding/zephyr_test/src/test.cpp:14:41: note: in expansion of macro ‘ARRAY_SIZE’
     bl_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), NULL, NULL);
                                         ^~~~~~~~~~
/home/zigarrre/coding/zephyr_test/src/test.cpp:14:68: error: expected ‘)’ before ‘;’ token
     bl_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), NULL, NULL);
                                                                    ^
/home/zigarrre/coding/zephyr_test/src/test.cpp:14:68: error: expected ‘)’ before ‘;’ token
/home/zigarrre/coding/zephyr_test/src/test.cpp:14:68: error: expected ‘)’ before ‘;’ token
/home/zigarrre/coding/zephyr_test/src/test.cpp: In function ‘int main()’:
/home/zigarrre/coding/zephyr_test/src/test.cpp:19:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }
 ^
make[2]: *** [CMakeFiles/app.dir/build.make:303: CMakeFiles/app.dir/src/test.cpp.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:228: CMakeFiles/app.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

# Full Examples

For additional, more complete code samples just use the various (Bluetooth) examples that Zephyr already provides and compile them as C++ by changing the extensions of the files as well as in the `CMakeFiles.txt` to `.cpp`.

I would suggest starting with `bluetooth/beacon` which is the most simple example and working our way up from there to more complex examples.

--- a/include/bluetooth/addr.h
+++ b/include/bluetooth/addr.h
@@ -32,12 +32,12 @@ typedef struct {
 	bt_addr_t a;
 } bt_addr_le_t;
 
-#define BT_ADDR_ANY     (&(bt_addr_t) { { 0, 0, 0, 0, 0, 0 } })
-#define BT_ADDR_NONE    (&(bt_addr_t) { \
-			 { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff } })
-#define BT_ADDR_LE_ANY  (&(bt_addr_le_t) { 0, { { 0, 0, 0, 0, 0, 0 } } })
-#define BT_ADDR_LE_NONE (&(bt_addr_le_t) { 0, \
-			 { { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff } } })
+#define BT_ADDR_ANY     ((bt_addr_t[]) { { { 0, 0, 0, 0, 0, 0 } } })
+#define BT_ADDR_NONE    ((bt_addr_t[]) { { \
+			 { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff } } })
+#define BT_ADDR_LE_ANY  ((bt_addr_le_t[]) { { 0, { { 0, 0, 0, 0, 0, 0 } } } })
+#define BT_ADDR_LE_NONE ((bt_addr_le_t[]) { { 0, \
+			 { { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff } } } })
 
 static inline int bt_addr_cmp(const bt_addr_t *a, const bt_addr_t *b)
 {

--- a/include/bluetooth/bluetooth.h
+++ b/include/bluetooth/bluetooth.h
@@ -321,11 +321,11 @@ struct bt_le_adv_param {
   * @param _int_max   Maximum advertising interval
   */
 #define BT_LE_ADV_PARAM(_options, _int_min, _int_max) \
-		(&(struct bt_le_adv_param) { \
+		((struct bt_le_adv_param[]) { { \
 			.options = (_options), \
 			.interval_min = (_int_min), \
 			.interval_max = (_int_max), \
-		 })
+		 } })
 
 #define BT_LE_ADV_CONN BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE, \
 				       BT_GAP_ADV_FAST_INT_MIN_2, \
@@ -452,12 +452,12 @@ struct bt_le_scan_param {
   * @param _window   Scan Window (N * 0.625 ms)
   */
 #define BT_LE_SCAN_PARAM(_type, _filter, _interval, _window) \
-		(&(struct bt_le_scan_param) { \
+		((struct bt_le_scan_param[]) { { \
 			.type = (_type), \
 			.filter_dup = (_filter), \
 			.interval = (_interval), \
 			.window = (_window), \
-		 })
+		 } })
 
 /** Helper macro to enable active scanning to discover new devices. */
 #define BT_LE_SCAN_ACTIVE BT_LE_SCAN_PARAM(BT_LE_SCAN_TYPE_ACTIVE, \

--- a/include/bluetooth/conn.h
+++ b/include/bluetooth/conn.h
@@ -46,12 +46,12 @@ struct bt_le_conn_param {
   * @param to       Supervision Timeout (N * 10 ms)
   */
 #define BT_LE_CONN_PARAM(int_min, int_max, lat, to) \
-	(&(struct bt_le_conn_param) { \
+	((struct bt_le_conn_param[]) { { \
 		.interval_min = (int_min), \
 		.interval_max = (int_max), \
 		.latency = (lat), \
 		.timeout = (to), \
-	 })
+	 } })
 
 /** Default LE connection parameters:
   *   Connection Interval: 30-50 ms
@@ -859,9 +859,9 @@ struct bt_br_conn_param {
   * @param role_switch True if role switch is allowed
   */
 #define BT_BR_CONN_PARAM(role_switch) \
-	(&(struct bt_br_conn_param) { \
+	((struct bt_br_conn_param[]) { { \
 		.allow_role_switch = (role_switch), \
-	 })
+	 } })
 
 /** Default BR/EDR connection parameters:
   *   Role switch allowed

--- a/include/bluetooth/gatt.h
+++ b/include/bluetooth/gatt.h
@@ -546,9 +546,9 @@ ssize_t bt_gatt_attr_read_chrc(struct bt_conn *conn,
 #define BT_GATT_CHARACTERISTIC(_uuid, _props, _perm, _read, _write, _value) \
 	BT_GATT_ATTRIBUTE(BT_UUID_GATT_CHRC, BT_GATT_PERM_READ,		\
 			  bt_gatt_attr_read_chrc, NULL,			\
-			  (&(struct bt_gatt_chrc) { .uuid = _uuid,	\
-						    .value_handle = 0U, \
-						    .properties = _props, })), \
+			  ((struct bt_gatt_chrc[]) { { .uuid = _uuid,	\
+						       .value_handle = 0U, \
+						       .properties = _props, } })), \
 	BT_GATT_ATTRIBUTE(_uuid, _perm, _read, _write, _value)
 
 #if IS_ENABLED(CONFIG_BT_SETTINGS_CCC_LAZY_LOADING)
@@ -660,8 +660,8 @@ ssize_t bt_gatt_attr_write_ccc(struct bt_conn *conn,
  *  @param _perm CCC access permissions.
  */
 #define BT_GATT_CCC(_changed, _perm)				\
-	BT_GATT_CCC_MANAGED((&(struct _bt_gatt_ccc)			\
-		BT_GATT_CCC_INITIALIZER(_changed, NULL, NULL)), _perm)
+	BT_GATT_CCC_MANAGED(((struct _bt_gatt_ccc[])			\
+		{BT_GATT_CCC_INITIALIZER(_changed, NULL, NULL)}), _perm)
 
 /** @brief Read Characteristic Extended Properties Attribute helper
  *

--- a/include/bluetooth/uuid.h
+++ b/include/bluetooth/uuid.h
@@ -69,11 +69,11 @@ struct bt_uuid_128 {
 }
 
 #define BT_UUID_DECLARE_16(value) \
-	((struct bt_uuid *) (&(struct bt_uuid_16) BT_UUID_INIT_16(value)))
+	((struct bt_uuid *) ((struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))
 #define BT_UUID_DECLARE_32(value) \
-	((struct bt_uuid *) (&(struct bt_uuid_32) BT_UUID_INIT_32(value)))
+	((struct bt_uuid *) ((struct bt_uuid_32[]) {BT_UUID_INIT_32(value)}))
 #define BT_UUID_DECLARE_128(value...) \
-	((struct bt_uuid *) (&(struct bt_uuid_128) BT_UUID_INIT_128(value)))
+	((struct bt_uuid *) ((struct bt_uuid_128[]) {BT_UUID_INIT_128(value)}))
 
 #define BT_UUID_16(__u) CONTAINER_OF(__u, struct bt_uuid_16, uuid)
 #define BT_UUID_32(__u) CONTAINER_OF(__u, struct bt_uuid_32, uuid)

--- a/include/data/json.h
+++ b/include/data/json.h
@@ -213,14 +213,14 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 		.offset = offsetof(struct_, field_name_), \
 		{ \
 			.array = { \
-				.element_descr = &(struct json_obj_descr) { \
+				.element_descr = (struct json_obj_descr[]) { { \
 					.align_shift = \
 						Z_ALIGN_SHIFT(struct_), \
 					.type = elem_type_, \
 					.offset = \
 						offsetof(struct_, \
 							 len_field_), \
-				}, \
+				} }, \
 				.n_elements = (max_len_), \
 			}, \
 		}, \
@@ -275,7 +275,7 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 		.offset = offsetof(struct_, field_name_), \
 		{ \
 			.array = { \
-				.element_descr = &(struct json_obj_descr) { \
+				.element_descr = (struct json_obj_descr[]) { { \
 					.align_shift = \
 						Z_ALIGN_SHIFT(struct_), \
 					.type = JSON_TOK_OBJECT_START, \
@@ -289,7 +289,7 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 							    elem_descr_len_, \
 						}, \
 					}, \
-				}, \
+				} }, \
 				.n_elements = (max_len_), \
 			}, \
 		}, \
@@ -353,7 +353,7 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 		.offset = offsetof(struct_, field_name_), \
 		{ \
 			.array = { \
-				.element_descr = &(struct json_obj_descr) { \
+				.element_descr = (struct json_obj_descr[]) { { \
 					.align_shift = \
 						Z_ALIGN_SHIFT(struct_), \
 					.type = JSON_TOK_LIST_START, \
@@ -367,7 +367,7 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 							    elem_descr_len_, \
 						}, \
 					}, \
-				}, \
+				} }, \
 				.n_elements = (max_len_), \
 			}, \
 		}, \
@@ -464,13 +464,13 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 		.offset = offsetof(struct_, struct_field_name_), \
 		{ \
 			.array = { \
-				.element_descr = &(struct json_obj_descr) { \
+				.element_descr = (struct json_obj_descr[]) { { \
 					.align_shift = \
 						Z_ALIGN_SHIFT(struct_), \
 					.type = elem_type_, \
 					.offset = offsetof(struct_, \
 							   len_field_), \
-				}, \
+				} }, \
 				.n_elements = (max_len_), \
 			}, \
 		}, \
@@ -534,7 +534,7 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 		.offset = offsetof(struct_, struct_field_name_), \
 		{ \
 			.array = { \
-				.element_descr = &(struct json_obj_descr) { \
+				.element_descr = (struct json_obj_descr[]) { { \
 					.align_shift = \
 						Z_ALIGN_SHIFT(struct_), \
 					.type = JSON_TOK_OBJECT_START, \
@@ -548,7 +548,7 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 							    elem_descr_len_, \
 						}, \
 					}, \
-				}, \
+				} }, \
 				.n_elements = (max_len_), \
 			}, \
 		}, \

